cpp/include/rapidsmpf/shuffler/chunk_batch.hpp

```diff
@@ -0,0 +1,390 @@
+            assert(
+                std::ptrdiff_t(metadata_buffer_->size())
+                >= metadata_offset + chunk_metadata_header_size
+            );
+
+            auto const* chunk_header =
+                reinterpret_cast<Chunk::MetadataMessageHeader const*>(
+                    metadata_buffer_->data() + metadata_offset
+                );
+            metadata_offset += chunk_metadata_header_size;
+
+            assert(
+                metadata_buffer_->size()
+                >= size_t(metadata_offset) + chunk_header->metadata_size
+            );
+
+            if (chunk_header->gpu_data_size > 0) {
+                assert(payload_data_);
+                assert(
+                    payload_data_->size
+                    >= size_t(payload_offset) + chunk_header->gpu_data_size
+                );
+            }
+
+            visitor(
+                chunk_header,
+                *metadata_buffer_,
+                metadata_offset,
+                *payload_data_,
+                payload_offset
+            );
+
+            metadata_offset += std::ptrdiff_t(chunk_header->metadata_size);
+            payload_offset += std::ptrdiff_t(chunk_header->gpu_data_size);
+        }
+    }
+
+    /**
+     * @brief Iterator to the beginning of the Chunk batch
+     *
+     * @param stream The stream to use for any memory allocations.
+     * @return Forward iterator to the beginning
+     */
+    [[nodiscard]] iterator begin(rmm::cuda_stream_view stream) const;
+
+    /**
+     * @brief Iterator to the end of the Chunk batch
+     *
+     * @param stream The stream to use for any memory allocations.
+     * @return Forward iterator to the end
+     */
+    [[nodiscard]] iterator end(rmm::cuda_stream_view stream) const;
+
+  private:
+    ChunkBatch(
+        std::unique_ptr<std::vector<uint8_t>> metadata,
+        std::unique_ptr<Buffer> payload_data
+    );
+
+    /// A buffer containing the BatchHeader, and metadata header and metadata of each
+    /// chunk. This buffer will not be null.
+    /// |BatchHeader|[[MetadataMessageHeader, Metadata], ...]|
+    ///
+    /// TODO: change the format to have the MetadataMessageHeaders at the front (after
+    /// BatchHeader), followed by the metadata. This will be a more cache efficient
+    /// traversal pattern.
+    std::unique_ptr<std::vector<uint8_t>> metadata_buffer_;
+
+    std::unique_ptr<Buffer>
+        payload_data_;  ///<  GPU data buffer associated with this batch of chunks. This
+                        ///<  buffer could be null if there were no chunks with payloads.
+
+    mutable std::unique_ptr<iterator>
+        cached_begin_iter_;  ///< cached begin iterator, as creating an iterator object is
+                             ///< relatively expensive. This allows calling `begin()`
+                             ///< trivially multiple times.
+};
+
+/**
+ * @brief Forward iterator of chunks in the chunk batch.
+ */
+class ChunkForwardIterator {
```

Comment:
I think I would prefer to implement `operator()[]` and `.at` on the `ChunkBatch` class so that we could just have a `RandomAccessIterator`. I think this might require a few changes in the metadata we store in a `ChunkBatch` (basically we need to keep the exclusive scan of the offsets around).